Added option to create keys under the EK in keygen example

tomoveu opened the pull request:
This PR adds creation of keys under the Endorsement Hierarchy. For this purpose, keygen can now create either:
- primary under Endorsement Hierarchy
or
- primary under Storage Hierarchy

Surprisingly, the creation of the attestation key fails with , although EK auth is empty(default) and set.
```
Creating new RSA key...
TPM2_Create key failed 303: TPM_RC_AUTH_UNAVAILABLE: The authValue or authPolicy is not available for selected entity
wolfTPM2_CreateKey failed

Failure 0x12f: TPM_RC_AUTH_UNAVAILABLE: The authValue or authPolicy is not available for selected entity
```

It could be that some auth slots get overwritten or maybe there is Admin flag set somewhere. To be investigated.

ps: This is needed to extend the Make/ActivateCredential examples 

Signed-off-by: Dimitar Tomov <dimi@wolfssl.com>

tomoveu commented:
This really bugged me out, so I debugged over the weekend. 

* EK requires an Endorsement auth that is in the form of TPM 2.0 Policy Session.
* It looks like after adding HMAC sessions, the use of Policy session might be impacted

wolfTPM2 wrappers may need modifications to enable keys under the Endorsement Hierarchy:

* Wrappers anticipate Password Auth at slot 0
* Authorization sessions should come before Password sessions 
* Even when Policy Session for EK auth is set, it is overwritten by the wrappers API

I thikn this can be fixed internally and without API implications.

Here is my WiP

tomoveu commented:
@dgarske I think the policy handling issue is fixed.

Remains one issue - the wrong Auth Area Size when Policy Session is used.

dgarske commented on examples/keygen/keygen.c:
```diff
@@ -108,7 +109,9 @@ int TPM2_Keygen_Example(void* userCtx, int argc, char *argv[])
     WOLFTPM2_DEV dev;
+    WOLFTPM2_KEY endorse; /* EK */
     WOLFTPM2_KEY storage; /* SRK */
+    WOLFTPM2_KEY *primary = &storage; /* Default option, see "-eh" in usage() */
```
The init can be removed here. It is always set in the args below.

tomoveu commented:
Latest commit fixes the issue with incorrect _auth area size_ field when multiple auth sessions are used.

Origin of the issue was the replacement of auth session data without updating the total auth area size field.

I am making further adjustments in the TPM2 command/packet handling.

tomoveu commented:
Issues fixed:
   * Workaround policy Session for EK auth
    * Added two new utility functions for TPM2 sessions
    * Added flag in WOLFTPM2_HANDLE to specify when policyAuth is used
I can now use AK under EK and will update the attestation example

tomoveu commented:
@dgarske please make a review before I rebase

This PR achieves way more than its goal, multiple improvements and new wrappers added in the process

dgarske commented on src/tpm2_wrap.c:
```diff
@@ -1661,6 +1751,33 @@ int wolfTPM2_RsaKey_TpmToWolf(WOLFTPM2_DEV* dev, WOLFTPM2_KEY* tpmKey,
+                             byte* pem, int* pemSz, byte* tempBuf, int tempSz)
+{
+    int rc = TPM_RC_FAILURE;
+#ifndef WOLFTPM2_NO_WOLFCRYPT
```
This also needs to check `WOLFSSL_DER_TO_PEM`.

dgarske commented on src/tpm2_wrap.c:
```diff
@@ -1726,6 +1843,30 @@ int wolfTPM2_RsaKey_WolfToTpm(WOLFTPM2_DEV* dev, RsaKey* wolfKey,
+                             byte *tempBuf, int tempSz, const char *filename)
+{
+    int rc = TPM_RC_FAILURE;
+#ifndef WOLFTPM2_NO_WOLFCRYPT
```
This also needs to check `defined(WOLFSSL_CERT_EXT) || defined(WOLFSSL_PUB_PEM_TO_DER) || defined(WOLFSSL_PEM_TO_DER)`

dgarske commented on wolftpm/tpm2.h:
```diff
@@ -3062,6 +3062,35 @@ WOLFTPM_API TPM_RC TPM2_SetSessionAuth(TPM2_AUTH_SESSION *session);
+    \sa TPM2_IsHmacSession
+    \sa TPM2_GetSessionAuthCount
+*/
+WOLFTPM_API int TPM2_isPolicySession(int sessionHandle);
```
Any luck making these macros? `#define TPM2_IsPolicySession(sessionHandle) ((sessionHandle) & 0x03000000)`

dgarske commented on wolftpm/tpm2.h:
```diff
@@ -3062,6 +3062,35 @@ WOLFTPM_API TPM_RC TPM2_SetSessionAuth(TPM2_AUTH_SESSION *session);
+    \sa TPM2_IsPolicySession
+    \sa TPM2_GetSessionAuthCount
+*/
+WOLFTPM_API int TPM2_isHmacSession(int sessionHandle);
```
Any luck making these macros? `#define TPM2_IsHmacSession(sessionHandle) ((sessionHandle) & 0x02000000)`

dgarske commented on src/tpm2_wrap.c:
```diff
@@ -1726,6 +1843,30 @@ int wolfTPM2_RsaKey_WolfToTpm(WOLFTPM2_DEV* dev, RsaKey* wolfKey,
 }
+
+int wolfTPM2_RsaKey_PemToTpm(WOLFTPM2_DEV* dev, WOLFTPM2_KEY* tpmKey,
+                             byte *tempBuf, int tempSz, const char *filename)
```
I would prefer a buffer PEM input, not filename. Use separate code to write to file. Use `wc_PubKeyPemToDer`

tomoveu replied:
My intention was to create a function that allows a user to import Private PEM Key into the TPM using this helper. I imagined a scenario, where a user wants to import his OpenSSH or OpenVPN key.

I think wc_PemPubKeyToDer allows conversion of both private and public keys, correct?

I made the change you requested. Perhaps I would have to rename the function to wolfTPM2_Rsa_Key_PubPemToTpm, so we leave space for future API of importing private PEM ?

dgarske commented on src/tpm2.c:
```diff
@@ -1068,6 +1097,51 @@ TPM_RC TPM2_Create(Create_In* in, Create_Out* out)
     return rc;
 }
 
+TPM_RC TPM2_CreateLoaded(CreateLoaded_In* in, CreateLoaded_Out* out)
```
Thanks for adding this missing TPM2_ API! I did not realize I had missed one.

dgarske commented on src/tpm2_wrap.c:
```diff
@@ -1027,6 +1061,62 @@ int wolfTPM2_CreateAndLoadKey(WOLFTPM2_DEV* dev, WOLFTPM2_KEY* key,
+    keyBlob->handle.symmetric = createLoadedOut.outPublic.publicArea.parameters.asymDetail.symmetric;
+
+    keyBlob->handle.hndl = createLoadedOut.objectHandle;
+    keyBlob->pub = createLoadedOut.outPublic;
```
Please use the new `wolfTPM2_CopyPub`, `wolfTPM2_CopyPriv` and `wolfTPM2_CopyName`

dgarske commented on src/tpm2.c:
```diff
@@ -517,6 +526,26 @@ int TPM2_GetSessionAuthCount(TPM2_CTX* ctx)
     return sessionCount;
 }
 
+int TPM2_isPolicySession(int sessionHandle)
```
See comment in header. Can we make these macros??

tomoveu commented:
@dgarske peer feedback taken, rebased on latest master

after rebasing activate credential gives TPM_RC_INTEGRITY error, so I am debugging

dgarske commented on src/tpm2_wrap.c:
```diff
@@ -1875,7 +1878,7 @@ int wolfTPM2_RsaKey_WolfToTpm(WOLFTPM2_DEV* dev, RsaKey* wolfKey,
 
 int wolfTPM2_RsaKey_PemToTpm(WOLFTPM2_DEV* dev, WOLFTPM2_KEY* tpmKey,
-                             byte *tempBuf, int tempSz, const char *filename)
+                             byte* pem, int pemSz, byte* tempBuf, int tempSz)
```
Yes `wolfTPM2_RsaKey_PubPemToTpm`?

tomoveu commented:
Fixed rebase issue, found out some wrap tests failed. investingating.

tomoveu commented:
@dgarske ready for merge

* Fixed wrap tests after rebase
* Added new unit tests
* Additional fixes from peer feeback
* Updated examples readme

dgarske commented:
@tomoveu please review my updates and let me know if they look good. Then we can merge.

tomoveu commented on src/tpm2_wrap.c:
```diff
@@ -1784,29 +1784,58 @@ int wolfTPM2_RsaKey_TpmToWolf(WOLFTPM2_DEV* dev, WOLFTPM2_KEY* tpmKey,
+            }
+        }
+        if (derSz > 0) {
+            derBuf = (byte*)XMALLOC(derSz, NULL, DYNAMIC_TYPE_TMP_BUFFER);
```
I think this is the first ever XMALLOC in wolfTPM. It is in the examples, but are we okay with starting to use HEAP?

We will have to update the product page.

dgarske replied:
It is not the first. Since this already requires wolfCrypt and the PEM code is already doing malloc it is fine.

tomoveu replied:
you are right. Good for merge then :)

tomoveu reviewed (commented):
Are we okay with using MALLOC for the very first time in wolfTPM?

Otherwise, code changes are great. Definitely good imporvements.
